dev-tools/mage/check.go

```diff
@@ -99,7 +99,7 @@ func GitDiffIndex() ([]string, error) {
 			return nil, errors.Wrap(err, "failed to dissect git diff-index output")
 		}
 
-		paths := strings.Split(m["paths"], "\t")
+		paths := strings.Split(m["paths"].(string), "\t")
```

Comment:
Why this change here?

Reply:
When I run **make fmt** I get following  error because d.Dissect() on line #97 now returns **map[string]interface{}**
`
~/go/src/github.com/elastic/beats/libbeat $ make fmt
github.com/elastic/beats/v7/dev-tools/mage
../dev-tools/mage/check.go:102:27: cannot use m["paths"] (type interface {}) as type string in argument to strings.Split: need type assertion
Error: error compiling magefiles
make: *** [add-headers] Error 1
`

Reply:
I ran the benchmark tests on **master** branch and **dissect_type_aware** branch. Most of the test cases degrade about 10%. Here is comparison of ns/op for master vs this feature


Test   case | New feature ns/op | Master ns/op | Diff
-- | -- | -- | --
BenchmarkDissect/When_all_the_defined_fields_are_captured_by_we_have_remaining_data-12 | 1253 | 1105 | 13%
BenchmarkDissect/Complex_stack_trace-12 | 1559 | 1359 | 15%
BenchmarkDissect/fails_when_delimiter_is_not_found_at_the_beginning_of_the_string-12 | 0.000022 | 0.000022 | 0%
BenchmarkDissect/fails_when_delimiter_is_not_found_after_the_key-12 | 0.000027 | 0.000026 | 4%
BenchmarkDissect/simple_dissect-12 | 852 | 815 | 5%
BenchmarkDissect/dissect_two_replacement-12 | 952 | 1066 | -11%
BenchmarkDissect/one_level_dissect_not_end_of_string-12 | 916 | 872 | 5%
BenchmarkDissect/one_level_dissect-12 | 886 | 832 | 6%
BenchmarkDissect/multiple_keys_dissect_end_of_string-12 | 953 | 903 | 6%
BenchmarkDissect/multiple_keys_not_end_of_string-12 | 1032 | 960 | 8%
BenchmarkDissect/simple_ordered-12 | 1189 | 1076 | 11%
BenchmarkDissect/simple_append-12 | 1160 | 1078 | 8%
BenchmarkDissect/indirect_field-12 | 1006 | 949 | 6%
BenchmarkDissect/skip_field-12 | 929 | 890 | 4%
BenchmarkDissect/named_skiped_field_with_indirect-12 | 1005 | 959 | 5%
BenchmarkDissect/pointer_field_with_indirect-12 | 997 | 935 | 7%
BenchmarkDissect/missing_fields-12 | 1303 | 1167 | 12%
BenchmarkDissect/ignore_right_padding-12 | 1281 | 1028 | 25%
BenchmarkDissect/padding_on_the_last_key_need_a_delimiter-12 | 1110 | 1040 | 7%
BenchmarkDissect/ignore_left_padding-12 | 1083 | 1027 | 5%
BenchmarkDissect/when_the_delimiters_contains_`{`_and_`}`-12 | 1060 | 1004 | 6%
BenchmarkDissect/simple_fixed_length-12 | 1039 | 956 | 9%
BenchmarkDissect/simple_ordered_and_fixed_length_field-12 | 1191 | 1101 | 8%
BenchmarkDissect/simple_padding_and_fixed_length_field-12 | 1227 | 1111 | 10%
BenchmarkDissect/mixed_pointer_and_indirect_and_fixed_length-12 | 1029 | 961 | 7%
BenchmarkDissect/fails_when_there_is_remaining_string_after_the_fixed-length_key-12 | 0.000052 | 0.000055 | -5%
BenchmarkDissect/fails_when_there_is_no_enough_string_for_the_fixed-length_key-12 | 0.000027 | 0.000026 | 4%
BenchmarkDissect/Regular_expression-12 | 422 | 421 | 0%
BenchmarkDissect/Larger_regular_expression-12 | 2794 | 2793 | 0%
BenchmarkDissect/regular_expression_to_match_end_of_line-12 | 586 | 571 | 3%